cpp/test/core/apply_tests.cpp

```diff
@@ -134,6 +141,210 @@ Bonded(uint16_t version = bond::v1)
+            _this << std::boolalpha << "struct " << *_qualifiedName << "\n{\n";
+        else
+            _structs = {};
+    }
+
+    template <typename T>
+    bool Base(const T& value) const
+    {
+        if (!_structs)
+            return true;
+
+        _this.seekp(-3, std::ios_base::end);
+        _this << " : " << StructName(value) << "\n{\n";
+        return false;
+    }
+
+    template <typename T>
+    bool Field(uint16_t id, const bond::Metadata& metadata, const T& value) const
+    {
+        if (!_structs)
+            return true;
+
+        _this << "    " << id << ": " << ToString(metadata.modifier) << ' ';
+        TypeName(value);
+        _this << ' ' << metadata.name;
+
+        if (metadata.default_value.nothing)
+            _this << " = nothing";
+        else
+            DefaultValue<typename bond::remove_bonded_value<T>::type>(metadata.default_value);
+
+        _this << ";\n";
+        return false;
+    }
+
+    template <typename T>
+    void Container(const T& value, std::size_t size) const
+    {
+        Container(size, value);
+    }
+
+    template <typename K, typename T>
+    void Container(const K& key, const T& value, std::size_t size) const
+    {
+        Container(size, key, value);
+    }
+
+    template <typename T>
+    bool UnknownField(std::uint16_t /*id*/, const T& /*value*/) const BOND_NOEXCEPT
+    {
+        return !_structs;
+    }
+
+    void UnknownEnd() const BOND_NOEXCEPT
+    {}
+
+    void End() const
+    {
+        if (!_structs)
+            return;
+
+        _this << "};\n";
+        _structs->emplace(*_qualifiedName, _this.str());
+    }
+
+private:
+    template <typename T>
+    const std::string& StructName(const T& value) const
+    {
+        IDLBuilder that{ *_structs };
+        Apply(that, value);
+        return *that._qualifiedName;
+    }
+
+    template <typename T, typename Reader, typename boost::enable_if<bond::is_basic_type<T> >::type* = nullptr>
+    void TypeName(const bond::value<T, Reader>& /*value*/) const
+    {
+        _this << bond::detail::type<typename std::conditional<std::is_enum<T>::value, std::int32_t, T>::type>::name();
+    }
+
+    template <typename T>
+    void TypeName(const T& value) const
+    {
+        _container = bond::GetTypeId(value);
+
+        if (_container == bond::BT_STRUCT)
+            _this << StructName(value);
+        else
+            Apply(*this, value);
+    }
+
+    template <typename... T>
+    void Container(std::size_t size, const T&... value) const
+    {
+        BOOST_VERIFY(size == 0);
+        assert(bond::BT_LIST <= _container && _container <= bond::BT_MAP);
+        assert((sizeof...(T) == 2) == (_container == bond::BT_MAP));
+
+        const char* labels[] = { "list", "set", "map" };
+        _this << labels[_container - bond::BT_LIST] << '<';
+        (void)std::initializer_list<int>{ (TypeName(value), _this << ", ", 0)... };
+        _this.seekp(-2, std::ios_base::end);
```

Comment:
Minor: like above, a comment for the -2 would be nice. 'Removing \", \" from after last type'?